README.txt
archetypes.kss Package Readme
=============================

Overview
--------

KSS (Kinetic Style Sheets) for Archetypes


docs/HISTORY.txt
Changelog
=========

1.7.4 (unreleased)
------------------

- Nothing changed yet.


1.7.3 (2013-08-14)
------------------

- On KSS Validation, pass REQUEST to validate method.
  Without it, it can happen that the language negociator chooses
  a different language for error messages than the regular form validation.
  This can result in error messages in different languages for kss
  and normal form validation.
  [do3cc]


1.7.2 (2012-05-25)
------------------


1.7.1 (2011-07-19)
------------------

- The change r12412 in 1.6.0a3 introduced a regression,
  the field inline validating error message was not shown to the user.
  The error message is not always the "required field" message, so it's
  problematic to not show the error message in all cases.
  The warning_only parameter is now set to False in the issueFieldError call so
  the error message appears again.
  [vincentfretin]

1.7.0 (2011-01-03)
------------------

- Use plone.uuid to look up content UUIDs.
  [toutpt, davisagli]

1.6.0 (2010-07-31)
------------------

- Calling it a 1.6.0 final, no changes since b1.
  [esteele]

1.6.0b1 (2010-05-01)
--------------------

- Check for inline editing status on an actual field in the content-core, not
  title which is now rendered by hand.
  [davisagli]

1.6.0a3 (2010-03-09)
--------------------

- Made error message be a warning instead before the form has been submitted,
  this causes less disruption in a form while editing.
  [joelburton]


1.6.0a2 (2010-03-09)
--------------------

- Fixed test failure in testVersionPreviewIsNotInlineEditable.
  [hannosch]

1.6.0a1 (2009-11-19)
--------------------

- Fixed validation test to work with plone.app.blob files.

- Updated reference to IATDocument from Products.ATContentTypes to the
  standard location.
  [hannosch]

1.4.3 (2008-11-02)
------------------

- Enable unlock on unload for inline edits.
  [alecm]

1.4.2 (2008-08-18)
------------------

- Using an adapter to determine whether inline editing is enabled, so
  the default behaviour can easily be overridden.
  [fschulze]

- Added check for site_property 'enable_inline_editing'. This is a global
  setting which can still be overwritten with kss_inline_editable.
  [fschulze]

- Adjusted deprecation warnings to point to Plone 4.0 instead of Plone 3.5
  since we changed the version numbering again.
  [hannosch]

1.4 (2008-04-21)
----------------

- test on global_inline_editable should happen only in
  getKssClassesInlineEditable()
  Fixed disabling of inline editing
  [fschulze]

- Some Archetypes-validators expect the request 
  as a keyword argument. Fixes #8005.
  [mborch]

1.4-rc1 (2008-03-25)
--------------------

- prepare for release
  [ree]

1.4-alpha1 (2008-03-09)
-----------------------

- prepare for release
  [ree]

1.2.6 (2008-01-03)
------------------

- Fixed i18n markup in fields_versioning.py.
  [hannosch]

1.2.5 (2008-01-03)
------------------

- Changed 'saveField' to not require 'value' explicitly
  but to take it from the request if missing.
  This was done to support the use of 
  saveField-kssSubmitForm: currentForm();
  which is needed for multi-valued form variables.
  [raphael]

1.2.4.1 (2007-12-07)
--------------------

- Handle aborted versioning due to exceedingly 
  large files.
  (Fixes http://dev.plone.org/plone/ticket/7236)
  [alecm]

1.2.4 (2007-12-06)
------------------

- Fix #7236: Handle unregistered objects with a version id 
  during inline edit.
  [alecm]

1.2.3 (2007-11-08)
------------------

- prepare for release
  [ree]

1.2.2 (2007-10-06)
------------------

- Prepare for release
  [ree]

1.2-1 (2007-09-10)
------------------

- UID refactoring for fields (inline editing) support
  [mborch]

- Make sure inline edit is not active in the versioning preview. 
  (fixes #6958). From now on, the kss_inline_editable
  variable can be defined to False in a template and this will
  cause the contained widgets to be not inline editable
  even though their condition would allow it.
  [ree]

- implement versioning work with inline edit (fixes #6938)
  [ree]

1.2 (2007-08-17)
----------------

- Fix bugs and test layer
  [ree, gotcha, jvloothuis, nouri]

1.2-rc2 (2007-07-27)
--------------------

- inline edit for archetypes widget in view mode, first steps
  [gotcha]

1.2-rc1.1 (2007-07-1a2)
-----------------------

- Prepare for release
  [ree]

1.2-rc1 (2007-07-08)
--------------------

- Cleaned up self.context usage. In browser views in Zope2 you need to do
  an aq_inner on self.context before using it.
  [hannosch]

1.2-beta2 (2007-04-30)
----------------------

- Prepare for release
  [ree]

1.2-beta1 (2007-03-04)
----------------------

- Prepare for release
  [ree]

1.2-alpha2 (2007-02-10)
-----------------------

- Preparation for release
  [ree]

- Initial package structure.
  [zopeskel]


archetypes/kss/tests/test_fields.py
# Copyright (c) 2006
# Authors:
#   Jean-Paul Ladage <j.ladage@zestsoftware.nl>
#
# This program is free software; you can redistribute it and/or modify
# it under the terms of the GNU General Public License version 2 as published
# by the Free Software Foundation.
#
# This program is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
# GNU General Public License for more details.
#
# You should have received a copy of the GNU General Public License
# along with this program; if not, write to the Free Software
# Foundation, Inc., 59 Temple Place - Suite 330, Boston, MA
# 02111-1307, USA.
#

import unittest
import re
from Products.PloneTestCase import PloneTestCase
from plone.app.kss.tests.kss_and_plone_layer import KSSAndPloneTestCase
from zope import component
from zope.component import adapter
from kss.core.interfaces import IKSSView
from archetypes.kss.interfaces import IVersionedFieldModifiedEvent
from kss.core.BeautifulSoup import BeautifulSoup

PloneTestCase.setupPloneSite()

@adapter(None, IKSSView, IVersionedFieldModifiedEvent)
def field_modified_handler(ob, view, event):
    ob._eventCaught = True

class FieldsViewTestCase(KSSAndPloneTestCase):

    def afterSetUp(self):
        PloneTestCase.PloneTestCase.afterSetUp(self)
        # commands will be rendered as data structures,
        self.setDebugRequest()
        self.loginAsPortalOwner()
        context = self.portal['front-page']
        # Set up a view 
        self.view = context.restrictedTraverse('saveField')

    # --
    # test the KSS methods
    # --

    def testReplaceField(self):
        self.view.context.changeSkin('Plone Default', self.view.request)
        result = self.view.replaceField('title', 'kss_generic_macros', 'title-field-view')
        self.assertEqual([(r['name'], r['selector'], r['selectorType'])
                             for r in result], [
            ('setStyle', '.portalMessage', 'css'),
            ('replaceInnerHTML', 'kssPortalMessage', 'htmlid'),
            ('setAttribute', 'kssPortalMessage', 'htmlid'),
            ('setStyle', 'kssPortalMessage', 'htmlid'),
            ('replaceHTML', 'parent-fieldname-title', 'htmlid'),
            ('focus', '#parent-fieldname-title .firstToFocus', '')
        ])

    def testReplaceWithView(self):
        self.view.context.changeSkin('Plone Default', self.view.request)
        result = self.view.replaceWithView('title', 'kss_generic_macros', 'title-field-view')
        self.assertEqual([(r['name'], r['selector'], r['selectorType'])
                             for r in result], [
             ('replaceHTML', 'parent-fieldname-title', 'htmlid'),
            ])

    # XXX this test should test without events, so it should stop events listening
    # but we have no more method for that
    def testSaveField(self):
        view = self.view
        result = view.saveField('title', {'title':'My Title'}, 
                                'kss_generic_macros', 'title-field-view')
        self.assertEqual([(r['name'], r['selector'], r['selectorType'])
                             for r in result], [
             ('replaceHTML', 'parent-fieldname-title', 'htmlid'),
            ])
        self.assertEqual('My Title', self.portal['front-page'].Title())
        res = view.saveField('description',
                             {'description':'Woot a funky description!'},
                             'kss_generic_macros', 'description-field-view')
        self.assertEqual('Woot a funky description!', self.portal['front-page'].Description())
    

    # XXX note how these tests are wwrong, obviously events are not listened in this setup
    def testSaveFieldWithEvents(self):
        view = self.view
        result = view.saveField('title', {'title':'My Title'}, 
                                'kss_generic_macros', 'title-field-view')
        self.assertEqual('My Title', self.portal['front-page'].Title())
        res = view.saveField('description',
                             {'description':'Woot a funky description!'},
                             'kss_generic_macros', 'description-field-view')
        self.assertEqual('Woot a funky description!', self.portal['front-page'].Description())


    def testSaveFieldWithValueFromRequest(self):
        view = self.view
        view.request.form['title'] = 'My Title'
        view.request.form['description'] = 'Woot a funky description!'
        result = view.saveField('title', None, 
                                'kss_generic_macros', 'title-field-view')
        self.assertEqual('My Title', self.portal['front-page'].Title())
        res = view.saveField('description',
                             None,
                             'kss_generic_macros', 'description-field-view')
        self.assertEqual('Woot a funky description!', 
                         self.portal['front-page'].Description())

    # XXX this test would only run, if events are really listened 
    # (which they are not) 
    def _XXX_testSaveFieldWithVersioning(self):
        view = self.view
        component.provideHandler(field_modified_handler)
        try:
            res = view.saveField('title', {'title':'My Title'}, 
                                    'kss_generic_macros', 'title-field-view')
            self.assert_(getattr(view.context, '_eventCaught', False))
            view.context._eventCaught = False
            res = view.saveField('description',
                                 {'description':'Woot a funky description!'},
                                 'kss_generic_macros', 'description-field-view')
            self.assert_(getattr(view.context, '_eventCaught', False))
        finally:
            sm = component.getSiteManager()
            sm.unregisterHandler(field_modified_handler)

    # XXX these tests with a provided target node id are a bit silly because the feature
    # is not currently in use in any templates -- we provide a target node id of
    # 'parent-fieldname-title' which is the default anyway -- but the tests make sure
    # this api extension work to some degree.
    
    def testReplaceFieldWithProvidedTargetNodeId(self):
        self.view.context.changeSkin('Plone Default', self.view.request)
        target = 'parent-fieldname-title'
        result = self.view.replaceField('title', 'kss_generic_macros', 'title-field-view', target=target)
        self.assertEqual([(r['name'], r['selector'], r['selectorType'])
                             for r in result], [
            ('setStyle', '.portalMessage', 'css'),
            ('replaceInnerHTML', 'kssPortalMessage', 'htmlid'),
            ('setAttribute', 'kssPortalMessage', 'htmlid'),
            ('setStyle', 'kssPortalMessage', 'htmlid'),
            ('replaceHTML', 'parent-fieldname-title', 'htmlid'),
            ('focus', '#parent-fieldname-title .firstToFocus', '')
        ])

    def testReplaceWithViewWithProvidedTargetNodeId(self):
        self.view.context.changeSkin('Plone Default', self.view.request)
        target = 'parent-fieldname-title'
        result = self.view.replaceWithView('title', 'kss_generic_macros', 'title-field-view', target=target)
        self.assertEqual([(r['name'], r['selector'], r['selectorType'])
                             for r in result], [
             ('replaceHTML', 'parent-fieldname-title', 'htmlid'),
            ])

    def testSaveFieldWithProvidedTargetNodeId(self):
        view = self.view
        target = 'parent-fieldname-title'
        result = view.saveField('title', {'title':'My Title'}, 
                                'kss_generic_macros', 'title-field-view', target=target)
        self.assertEqual([(r['name'], r['selector'], r['selectorType'])
                             for r in result], [
             ('replaceHTML', 'parent-fieldname-title', 'htmlid'),
            ])
        self.assertEqual('My Title', self.portal['front-page'].Title())

    def testReplaceFieldWithProvidedContext(self):
        # set the global context to /news
        context = self.portal['news']
        view = context.restrictedTraverse('saveField')
        context.changeSkin('Plone Default', view.request)

        frontpage_uid = self.portal['front-page'].UID()
        result = view.replaceField('title', 'kss_generic_macros', 'title-field-view', uid=frontpage_uid)

        self.assertEqual([(r['name'], r['selector'], r['selectorType'])
                             for r in result], [
            ('setStyle', '.portalMessage', 'css'),
            ('replaceInnerHTML', 'kssPortalMessage', 'htmlid'),
            ('setAttribute', 'kssPortalMessage', 'htmlid'),
            ('setStyle', 'kssPortalMessage', 'htmlid'),
            ('replaceHTML', 'parent-fieldname-title', 'htmlid'),
            ('focus', '#parent-fieldname-title .firstToFocus', '')
        ])

        # make sure we've got the right context:
        replaceHTML = ''.join([r['params'].get('html', '') for r in result])
        self.assertEqual(u"Welcome to Plone" in replaceHTML, True)       

    def testReplaceWithViewWithProvidedContext(self):
        # set the global context to /news
        context = self.portal['news']
        view = context.restrictedTraverse('saveField')
        context.changeSkin('Plone Default', view.request)

        frontpage_uid = self.portal['front-page'].UID()
        result = view.replaceWithView('title', 'kss_generic_macros', 'title-field-view', uid=frontpage_uid)

        self.assertEqual([(r['name'], r['selector'], r['selectorType'])
                             for r in result], [
             ('replaceHTML', 'parent-fieldname-title', 'htmlid'),
            ])

        # make sure we've got the right context:
        replaceHTML = ''.join([r['params'].get('html', '') for r in result])
        self.assertEqual(u"Welcome to Plone" in replaceHTML, True)       

    def testSaveFieldWithProvidedContext(self):
        # set the global context to /news
        context = self.portal['news']
        view = context.restrictedTraverse('saveField')
        context.changeSkin('Plone Default', view.request)

        frontpage_uid = self.portal['front-page'].UID()
        result = view.saveField('title', {'title': 'My Title'},
                                'kss_generic_macros', 'title-field-view', uid=frontpage_uid)

        self.assertEqual('My Title', self.portal['front-page'].Title())

    
    def testMarkerInATField(self):
        # writeable
        view = self.portal['front-page'].restrictedTraverse('kss_field_decorator_view')
        result = view.getKssClasses('title')
        self.assertEqual(result, ' kssattr-atfieldname-title')
        result = view.getKssClasses('title', 'template')
        self.assertEqual(result, ' kssattr-atfieldname-title kssattr-templateId-template')
        result = view.getKssClasses('title', 'template', 'macro')
        self.assertEqual(result, ' kssattr-atfieldname-title kssattr-templateId-template kssattr-macro-macro')
        self.logout()
        result = view.getKssClasses('title')
        # not writeable
        self.assertEqual(result, '')

    def testMarkerInATFieldInlineEditable(self):
        # writeable
        view = self.portal['front-page'].restrictedTraverse('kss_field_decorator_view')
        result = view.getKssClassesInlineEditable('title', 'template')
        self.assertEqual(result, ' kssattr-atfieldname-title kssattr-templateId-template inlineEditable')
        result = view.getKssClassesInlineEditable('title', 'template', 'macro')
        self.assertEqual(result, ' kssattr-atfieldname-title kssattr-templateId-template kssattr-macro-macro inlineEditable')
        self.logout()
        # not writeable
        result = view.getKssClassesInlineEditable('title', 'template')
        self.assertEqual(result, '')

    def testMarkerInNonATField(self):
        # portal root is not an AT object
        view = self.portal.restrictedTraverse('kss_field_decorator_view')
        result = view.getKssClasses('title')
        # not writeable
        self.assertEqual(result, '')

    def testVersionPreviewIsNotInlineEditable(self):
        """If the kss_inline_editable variable is defined to False
        in a page template, all the fields will be globally prohibited 
        to be editable. This works via the getKssClasses method.
        Similarly, is suppress_preview is set to true, inline
        editing is prohibited. This is set from CMFEditions, in the
        versions_history_form.

        In this test we check that the versions history is not
        inline editable at all.
        """
        obj = self.portal['front-page']
        # Make sure we actually have a revision
        pr = self.portal.portal_repository
        pr.save(obj)
        # Render versions history of the front page
        obj.REQUEST.form['version_id'] = '0'
        rendered = obj.versions_history_form()
        soup = BeautifulSoup(rendered)
        # check that inline edit is not active, by looking at body text
        tag = soup.find(id=re.compile('parent-fieldname-text-*'))
        klass = tag['class']
        # ... and now see we are really not inline editable:
        self.assert_('inlineEditable' not in klass)
        # make sure the rest is still there or instant validation and
        # possibly other stuff will fail
        self.assert_('kssattr-templateId-' in klass)
        self.assert_('kssattr-macro-' in klass)

def test_suite():
    return unittest.TestSuite((
        unittest.makeSuite(FieldsViewTestCase),
        ))


archetypes/kss/fields.py
# -*- coding: utf-8 -*-
# Copyright (c) 2006
# Authors:
#   Jean-Paul Ladage <j.ladage@zestsoftware.nl>, jladage
#
# This program is free software; you can redistribute it and/or modify
# it under the terms of the GNU General Public License version 2 as published
# by the Free Software Foundation.
#
# This program is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
# GNU General Public License for more details.
#
# You should have received a copy of the GNU General Public License
# along with this program; if not, write to the Free Software
# Foundation, Inc., 59 Temple Place - Suite 330, Boston, MA
# 02111-1307, USA.
#
#from zope.component import getMultiAdapter
#from zope.viewlet.interfaces import IViewletManager

from plone.app.kss.plonekssview import PloneKSSView
from plone.app.kss.interfaces import IPloneKSSView
from plone.app.kss.interfaces import IPortalObject

from plone.locking.interfaces import ILockable

from zope.interface import implements
from zope import lifecycleevent, event
from zope.publisher.interfaces.browser import IBrowserView

from Acquisition import aq_inner, aq_base
from Products.Archetypes.event import ObjectEditedEvent
from Products.Five.browser import BrowserView
from Products.Five.browser.pagetemplatefile import ZopeTwoPageTemplateFile
from Products.CMFCore.utils import getToolByName

from zope.deprecation import deprecated

import events
from utils import get_econtext

missing_uid_deprecation = \
"This view does not provide a KSS instance UID as required. Falling back to "
"the global context on inline-editing will be removed in Plone 4.0. Please "
"update your templates."

class FieldsView(PloneKSSView):
    
    implements(IPloneKSSView)

    ## KSS methods
   
    view_field_wrapper = ZopeTwoPageTemplateFile('browser/view_field_wrapper.pt')
    edit_field_wrapper = ZopeTwoPageTemplateFile('browser/edit_field_wrapper.pt')

    def renderViewField(self, fieldname, templateId, macro, uid=None):
        """
        renders the macro coming from the view template
        """

        context = self._getFieldContext(uid)
        template = self.getTemplate(templateId, context)

        viewMacro = template.macros[macro]
        res = self.view_field_wrapper(viewMacro=viewMacro,
                                      context=context,
                                      templateId=templateId,
                                      fieldName=fieldname)
        return res

    def getTemplate(self, templateId, context=None):
        """
        traverse/search template
        """

        if not context:
            context = self.context

        template = context.restrictedTraverse(templateId)
        
        if IBrowserView.providedBy(template):
            view = template
            for attr in ('index', 'template', '__call__'):
                template = getattr(view, attr, None)
                if template is not None and hasattr(template, 'macros'):
                    break
            if template is None:
                raise KeyError("Unable to find template for view %s" % templateId)
        return template


    def renderEditField(self, fieldname, templateId, macro, uid=None):
        """
        renders the edit widget inside the macro coming from the view template
        """

        context = self._getFieldContext(uid)
        template = self.getTemplate(templateId, context)
        containingMacro = template.macros[macro]
        fieldname = fieldname.split('archetypes-fieldname-')[-1]
        field = context.getField(fieldname)
        widget = field.widget
        widgetMacro = widget('edit', context)
        
        res = self.edit_field_wrapper(containingMacro=containingMacro,
                                      widgetMacro=widgetMacro,
                                      field=field, instance=context,
                                      mode='edit',
                                      templateId=templateId)

        return res


    # XXX XXX TODO for 3.0.1: see if we really need edit at all or we can remoce it (ree)

    def replaceField(self, fieldname, templateId, macro, uid=None, target=None, edit=False):
        """
        kss commands to replace the field value by the edit widget

        The edit parameter may be used if we are coming from something else
        than an edit view.
        """
        ksscore = self.getCommandSet('core')
        zopecommands = self.getCommandSet('zope')
        plonecommands = self.getCommandSet('plone')

        instance = self._getFieldContext(uid)        

        if edit:
            locking = ILockable(instance, None)
            if locking:
                if not locking.can_safely_unlock():
                    selector = ksscore.getHtmlIdSelector('plone-lock-status')
                    zopecommands.refreshViewlet(selector,
                                                'plone.abovecontent',
                                                'plone.lockinfo')
                    plonecommands.refreshContentMenu()

                    return self.render()
                else: # were are locking the content
                    locking.lock()

        plonecommands.issuePortalMessage('')

        html = self.renderEditField(fieldname, templateId, macro, uid)
        html = html.strip()

        field_id = target or "parent-fieldname-%s" % fieldname
        ksscore.replaceHTML(ksscore.getHtmlIdSelector(field_id), html)
        ksscore.focus("#%s .firstToFocus" % field_id)

        return self.render()

    def replaceWithView(self, fieldname, templateId, macro, uid=None, target=None, edit=False):
        """
        kss commands to replace the edit widget by the field view
        """

        ksscore = self.getCommandSet('core')
        
        instance = self._getFieldContext(uid)        
        locking = ILockable(instance, None)
        if locking and locking.can_safely_unlock():
            locking.unlock()

        html = self.renderViewField(fieldname, templateId, macro, uid)
        html = html.strip()

        field_id = target or "parent-fieldname-%s" % fieldname
        ksscore.replaceHTML(ksscore.getHtmlIdSelector(field_id), html)

        return self.render()

    def saveField(self, fieldname, value=None, templateId=None, macro=None, uid=None, target=None):
        """
        This method saves the current value to the field. and returns the rendered
        view macro.
        """
        # We receive a dict or nothing in value.
        #
        if value is None:
            value = self.request.form.copy()
        instance = self._getFieldContext(uid)        
        field = instance.getField(fieldname)
        value, kwargs = field.widget.process_form(instance, field, value)
        error = field.validate(value, instance, {}, REQUEST=self.request)
        if not error and field.writeable(instance):
            setField = field.getMutator(instance)
            setField(value, **kwargs)

            # send event that will invoke versioning
            events.fieldsModified(instance, fieldname)

            instance.reindexObject() #XXX: Temp workaround, should be gone in AT 1.5

            descriptor = lifecycleevent.Attributes(IPortalObject, fieldname)
            event.notify(ObjectEditedEvent(instance, descriptor))
            
            return self.replaceWithView(fieldname, templateId, macro, uid, target)
        else:
            if not error:
                # XXX This should not actually happen...
                error = 'Field is not writeable.'
            # Send back the validation error
            self.getCommandSet('atvalidation').issueFieldError(fieldname, error)
            return self.render()

    def _getFieldContext(self, uid):
        if uid is not None:
            rc = getToolByName(aq_inner(self.context), 'reference_catalog')
            return rc.lookupObject(uid)
        else:
            deprecated(FieldsView, missing_uid_deprecation)
            return aq_inner(self.context)
        
class ATDocumentFieldsView(FieldsView):

    def isTableOfContentsEnabled(self):
        getTableContents = getattr(self.context, 'getTableContents', None)
        result = False
        if getTableContents is not None:
            result = getTableContents()
        return result

    def replaceField(self, fieldname, templateId, macro, uid=None, target=None, edit=False):
        if fieldname == "text" and self.isTableOfContentsEnabled():  
            self.getCommandSet('core').setStyle("#document-toc", name="display", value="none")
        FieldsView.replaceField(self, fieldname, templateId, macro, uid=uid, target=target, edit=edit)
        return self.render()

    def replaceWithView(self, fieldname, templateId, macro, uid=None, target=None, edit=False):
        FieldsView.replaceWithView(self, fieldname, templateId, macro, uid=uid, target=target, edit=edit)
        if fieldname == "text" and self.isTableOfContentsEnabled(): 
            self.getCommandSet('core').setStyle("#document-toc", name="display", value="block")
            self.getCommandSet('plone-legacy').createTableOfContents()
        return self.render()
    
    def saveField(self, fieldname, value=None, templateId=None, macro=None, uid=None, target=None):
        FieldsView.saveField(self, fieldname,
                value = value,
                templateId = templateId,
                macro = macro,
                uid = uid,
                target = target,
                )
        if fieldname == "text" and self.isTableOfContentsEnabled(): 
            self.getCommandSet('plone-legacy').createTableOfContents() 
            #manager = getMultiAdapter((self.context, self.request, self),
            #                          IViewletManager,
            #                          name='plone.abovecontentbody')
            #self.getCommandSet('refreshviewlet').refreshViewlet('document-toc',
            #                                                    manager,
            #                                                    'plone.tableofcontents')
        return self.render()

# --
# (Non-ajax) browser view for decorating the field
# --

class ATFieldDecoratorView(BrowserView):

    def getKssUIDClass(self):
        """
        This method generates a class-name from the current context UID.
        """
        uid = aq_inner(self.context).UID()
        
        return "kssattr-atuid-%s" % uid

    def _global_kss_inline_editable(self):
        """With a nasty although not unusual hack, we reach
        out to the caller template, and examine the global
        tal variable kss_inline_editable. If it is defined,
        and if it is defined to false, we will prohibit
        inline editing everywhere in the template.
        We apply this 'magic' because the signature to getKssClasses
        is already too complex, and it would be undesirable to
        complicate it some more.
        """
        econtext = get_econtext()
        if econtext is None:
            # tests, probably
            return True
        # kss_inline_editable can be set to false in a template, and this
        # will prohibit inline editing in the page
        kss_inline_editable = econtext.vars.get('kss_inline_editable', None)
        # check the setting in site properties
        context = aq_inner(self.context)
        portal_properties = getToolByName(context, 'portal_properties')
        enable_inline_editing = None
        if getattr(aq_base(portal_properties), 'site_properties', None) is not None:
            site_properties = portal_properties.site_properties
            if getattr(aq_base(site_properties), 'enable_inline_editing', None) is not None:
                enable_inline_editing = site_properties.enable_inline_editing
        # If none of these is set, we enable inline editing. The global
        # site_property may be overwritten by the kss_inline_editable variable
        if kss_inline_editable is None:
            inline_editable = enable_inline_editing
        else:
            inline_editable = kss_inline_editable
        if inline_editable is None:
            inline_editable = True
        # In addition we also check suppress_preview.
        # suppress_preview is set by CMFEditions, when version preview is shown
        # This means inline editing should be disabled globally
        suppress_preview = econtext.vars.get('suppress_preview', False)
        return inline_editable and not suppress_preview

    def getKssClasses(self, fieldname, templateId=None, macro=None, target=None):
        context = aq_inner(self.context)
        field = context.getField(fieldname)
        # field can be None when widgets are used without fields
        # check whether field is valid
        if field is not None and field.writeable(context):
            classstring = ' kssattr-atfieldname-%s' % fieldname
            if templateId is not None:
                classstring += ' kssattr-templateId-%s' % templateId
            if macro is not None:
                classstring += ' kssattr-macro-%s' % macro
            if target is not None:
                classstring += ' kssattr-target-%s' % target
            # XXX commented out to avoid macro showing up twice
            # not removed since it might be needed in a use case I forgot about
            # __gotcha
            #else:
            #    classstring += ' kssattr-macro-%s-field-view' % fieldname
        else:
            classstring = ''
        return classstring
    
    def getKssClassesInlineEditable(self, fieldname, templateId, macro=None, target=None):
        classstring = self.getKssClasses(fieldname, templateId, macro, target)
        global_kss_inline_editable = self._global_kss_inline_editable()
        if global_kss_inline_editable and classstring:
            classstring += ' inlineEditable'
        return classstring
